apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPoolImpl.cs

```diff
@@ -0,0 +1,235 @@
+        {
+            Logger.Debug(() => "Attempting to acquire a session from the pool.");
+            SessionPoolEntry sessionEntry;
+            if (!TryPop(options, out sessionEntry))
+            {
+                Logger.Debug(() => "Attempting to acquire a session from the pool failed - creating a new instance.");
+                Stopwatch sw = null;
+                if (Logger.LogPerformanceTraces)
+                {
+                    sw = new Stopwatch();
+                    sw.Start();
+                }
+                //create a new session, blocking or throwing if at the limit.
+                var result = await Key.Client.CreateSessionAsync(new DatabaseName(Key.Project, Key.Instance, Key.Database), cancellationToken).ConfigureAwait(false);
+                if (sw != null)
+                {
+                    Logger.LogPerformanceCounterFn("Session.CreateTime", x => sw.ElapsedMilliseconds);
+                }
+                return result;
+            }
+            MarkUsed();
+            //note that the evict task will only actually delete the session if it was able to remove it from the pool.
+            //at this point, this is not possible because we removed it from the pool, so even if the task completes (which
+            // is possible due to a race), it will see that the session isn't pooled and cancel out.
+            sessionEntry.EvictTaskCancellationSource.Cancel();
+            return sessionEntry.Session;
+        }
+
+        public Session AcquireEvictionCandidate()
+        {
+            SessionPoolEntry sessionEntry = default(SessionPoolEntry);
+            lock (_sessionMruStack)
+            {
+                if (_sessionMruStack.Count > 0)
+                {
+                    sessionEntry = _sessionMruStack[_sessionMruStack.Count - 1];
+                    _sessionMruStack.RemoveAt(_sessionMruStack.Count - 1);
+                    Interlocked.Decrement(ref s_activeSessionsPooled);
+                    LogSessionsPooled();
+                }
+            }
+            if (sessionEntry.Session == null || sessionEntry.EvictTaskCancellationSource == null)
+            {
+                return null;
+            }
+            sessionEntry.EvictTaskCancellationSource.Cancel();
+            return sessionEntry.Session;
+        }
+
+        private void MarkUsed()
+        {
+            _lastAccessTime = Environment.TickCount;
+        }
+
+        public void ReleaseSessionToPool(SpannerClient client, Session session)
+        {
+            Logger.Debug(() => "Placing session back into the pool and starting the evict timer.");
+            //start evict timer.
+            SessionPoolEntry entry = new SessionPoolEntry(session, new CancellationTokenSource());
+
+            if (SessionPool.UseTransactionWarming)
+            {
+                Task.Run(() => client.PreWarmTransactionAsync(entry.Session));
```

Comment:
It all runs through the same code path here (to keep all paths the same) until it gets to actually filling in transaction details in TransactionPool.PreWarmAsync -- there you will see this code:
                // for now, we only prewarm readwrite transactions because the read transaction semantics are usually
                // dependent on the time the transaction begins.
                if (options?.ModeCase == TransactionOptions.ModeOneofCase.ReadWrite)
                {
                    Logger.Debug(
                        () => $"Pre-warming session transaction state. Mode={options.ModeCase}");
                    ActiveTransactionOptions = options;
                    PreWarmTask = Task.Run(() => CreateTransactionImplAsync(session));
                }
                else
                {
                    MarkTransactionUsed();
                }